### PR TITLE
Refactor connect_server_client and ssh tunnel logic to be more clear and reusable.

### PR DESCRIPTION
**Pull request by rohinb2**
reusable.

### CODE REVIEW

File: runhouse/resources/hardware/utils.py [hunk 1]
```diff
@@ -46,7 +46,7 @@
 # TODO: Move the following two functions into a networking module
 def get_open_tunnel(address: str, ssh_port: str):
```
Comment (jlewitt1) [inline runhouse/resources/hardware/utils.py:47]: shouldn't this be an int?

Comment (rohinb2) [reply runhouse/resources/hardware/utils.py:47]: yes good catch

File: runhouse/resources/hardware/cluster.py [hunk 2]
```diff
@@ -394,51 +391,46 @@ def connect_server_client(self, force_reconnect=False):
...
+                host=LOCALHOST,
                 port=self.client_port,
                 auth=auth,
-                cert_path=cert_path,
-                use_https=use_https,
             )
+
         else:
+            # Case 2: We're making a direct connection to the server, either via HTTP or HTTPS
+            assert self.server_connection_type in [
```
[the remark this answers is not in the document]
Comment (rohinb2) [reply runhouse/resources/hardware/cluster.py:426]: ye will do

File: runhouse/resources/hardware/utils.py [hunk 3]
```diff
@@ -448,3 +447,127 @@ def rsync(
...
+
+
+def ssh_tunnel(
+    address: str,
+    ssh_creds: Dict,
+    local_port: int,
+    ssh_port: int = 22,
+    remote_port: Optional[int] = None,
+    num_ports_to_try: int = 0,
+) -> Tuple[SSHTunnelForwarder, int]:
```
Comment (jlewitt1) [inline runhouse/resources/hardware/utils.py:475]: only returns the `SSHTunnelForwarder` now?

File: runhouse/resources/hardware/utils.py [hunk 4]
```diff
@@ -448,3 +447,127 @@ def rsync(
...
+    while not connected:
+        try:
+            if num_ports_to_try < 0:
+                raise Exception(
+                    f"Failed to create SSH tunnel after {num_ports_to_try} attempts"
+                )
+
+            if ssh_creds.get("ssh_proxy_command"):
+                # Start a tunnel using self.run in a thread, instead of ssh_tunnel
+                ssh_credentials = copy.copy(ssh_creds)
```
Comment (jlewitt1) [inline runhouse/resources/hardware/utils.py:528]: missing `copy` import?

**Pull request merged**